geopandas/tools/clip.py

```diff
@@ -0,0 +1,252 @@
+        resulting in multiple geometry types or GeometryCollections.
+        If False, return all resulting geometries (potentially mixed-types).
+
+    Returns
+    -------
+    GeoDataFrame or GeoSeries
+         Vector data (points, lines, polygons) from `gdf` clipped to
+         polygon boundary from mask.
+
+    Examples
+    --------
+    Clip points (global cities) with a polygon (the South American continent):
+
+    >>> import geopandas
+    >>> path =
+    >>> world = geopandas.read_file(
+    ...     geopandas.datasets.get_path('naturalearth_lowres'))
+    >>> south_america = world[world['continent'] == "South America"]
+    >>> capitals = geopandas.read_file(
+    ...     geopandas.datasets.get_path('naturalearth_cities'))
+    >>> capitals.shape
+    (202, 2)
+    >>> sa_capitals = geopandas.clip(capitals, south_america)
+    >>> sa_capitals.shape
+    (12, 2)
+    """
+    if not isinstance(gdf, (GeoDataFrame, GeoSeries)):
+        raise TypeError(
+            "'gdf' should be GeoDataFrame or GeoSeries, got {}".format(type(gdf))
+        )
+
+    if not isinstance(mask, (GeoDataFrame, GeoSeries, Polygon, MultiPolygon)):
+        raise TypeError(
+            "'mask' should be GeoDataFrame, GeoSeries or"
+            "(Multi)Polygon, got {}".format(type(gdf))
+        )
+
+    if isinstance(mask, (GeoDataFrame, GeoSeries)):
+        box_mask = mask.total_bounds
+    else:
+        box_mask = mask.bounds
+    box_gdf = gdf.total_bounds
+    if not (
+        ((box_mask[0] <= box_gdf[2]) and (box_gdf[0] <= box_mask[2]))
+        and ((box_mask[1] <= box_gdf[3]) and (box_gdf[1] <= box_mask[3]))
+    ):
+        return GeoDataFrame(columns=gdf.columns, crs=gdf.crs)
+
+    if isinstance(mask, (GeoDataFrame, GeoSeries)):
+        poly = mask.geometry.unary_union
+    else:
+        poly = mask
+
+    geom_types = gdf.geometry.type
+    poly_idx = np.asarray((geom_types == "Polygon") | (geom_types == "MultiPolygon"))
+    line_idx = np.asarray(
+        (geom_types == "LineString")
+        | (geom_types == "LinearRing")
+        | (geom_types == "MultiLineString")
+    )
+    point_idx = np.asarray((geom_types == "Point") | (geom_types == "MultiPoint"))
+    geomcoll_idx = np.asarray((geom_types == "GeometryCollection"))
+
+    if point_idx.any():
+        point_gdf = _clip_points(gdf[point_idx], poly)
+    else:
+        point_gdf = None
+
+    if poly_idx.any():
+        poly_gdf = _clip_line_poly(gdf[poly_idx], poly)
+    else:
+        poly_gdf = None
+
+    if line_idx.any():
+        line_gdf = _clip_line_poly(gdf[line_idx], poly)
+    else:
+        line_gdf = None
+
+    if geomcoll_idx.any():
+        geomcoll_gdf = _clip_line_poly(gdf[geomcoll_idx], poly)
+    else:
+        geomcoll_gdf = None
+
+    order = pd.Series(range(len(gdf)), index=gdf.index)
+    concat = pd.concat([point_gdf, line_gdf, poly_gdf, geomcoll_gdf])
+
+    geomcoll_concat = (concat.geom_type == "GeometryCollection").any()
```

Comment:
I think you can move this calculation (and the two on the lines below) inside the `if keep_geom_type`, because in the default case of False, we don't need to calculate those?